ballerina/channel_csv_io.bal

```diff
@@ -40,30 +40,60 @@ Error {
     return (check getReadableCSVChannel(readableChannel, 0)).csvStream();
 }
 
-isolated function channelWriteCsv(WritableChannel writableChannel, string[][]|map<anydata>[] content) returns Error? {
-    WritableCSVChannel csvChannel = check getWritableCSVChannel(writableChannel);
-    if content is string[][] {
-        foreach string[] r in content {
+isolated function channelWriteCsv(string path, FileWriteOption userProvidedOption, string[][]|map<anydata>[] contentToWrite) returns Error? {
+    WritableCSVChannel csvChannel;
+    FileWriteOption option = userProvidedOption;
+    if contentToWrite is string[][] {
+        csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        foreach string[] r in contentToWrite {
             Error? writeResult = csvChannel.write(r);
             if writeResult is Error {
                 check csvChannel.close();
                 return writeResult;
             }
         }
-    } else if content is map<anydata>[] {
-        foreach map<anydata> row in content {
+        check csvChannel.close();
+    } else if contentToWrite is map<anydata>[] {
+        string[] headers = [];
+        if contentToWrite.length() == 0 {
+            return;
+        }
+        headers = contentToWrite[0].keys();
+        if userProvidedOption == APPEND {
+            string[] headersFromCSV = check readHeadersFromCsvFile(path);
+            if headersFromCSV.length() > 0 {
+                headersFromCSV = check validateCsvHeaders(headersFromCSV, headers);
+                if (headersFromCSV.length() > 0) {
+                    headers = headersFromCSV;
+                    option = APPEND;
+                } else {
+                    option = OVERWRITE;
+                }
+            } else {
+                option = OVERWRITE;
+            }
+        }
+        csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        if option == OVERWRITE {
+            Error? headerWriteResult = csvChannel.write(headers);
+            if headerWriteResult is Error {
+                check csvChannel.close();
+                return headerWriteResult;
+            }
+        }
+        foreach map<anydata> row in contentToWrite {
             string[] sValues = [];
-            foreach [string, anydata] [_, value] in row.entries() {
-                sValues.push(value.toString());
+            foreach string header in headers {
+                sValues.push(row.get(header).toString());
             }
             Error? writeResult = csvChannel.write(sValues);
             if writeResult is Error {
                 check csvChannel.close();
                 return writeResult;
             }
         }
+        check csvChannel.close();
```

Comment:
Let's move write and append into different functions and call them. we can create the channel inside the function

Reply:
There is a code duplicate, but it is okay. it is more readable IMO